hedera-node/hedera-consensus-service-impl/src/main/java/com/hedera/node/app/service/consensus/impl/handlers/ConsensusSubmitMessageHandler.java

```diff
@@ -68,22 +89,162 @@ public void preHandle(@NonNull final PreHandleContext context, @NonNull Readable
+     * @param config the {@link ConsensusServiceConfig} for the active transaction
      * @param recordBuilder the {@link ConsensusSubmitMessageRecordBuilder} for the active transaction
      * @throws NullPointerException if one of the arguments is {@code null}
      */
     public void handle(
             @NonNull final HandleContext handleContext,
-            @NonNull final ConsensusSubmitMessageTransactionBody submitMessage,
-            @NonNull final ConsensusServiceConfig consensusServiceConfig,
-            @NonNull final ConsensusSubmitMessageRecordBuilder recordBuilder) {
-        throw new UnsupportedOperationException("Not implemented");
+            @NonNull final TransactionBody txn,
+            @NonNull final ConsensusServiceConfig config,
+            @NonNull final ConsensusSubmitMessageRecordBuilder recordBuilder,
+            @NonNull final WritableTopicStore topicStore) {
+        requireNonNull(handleContext);
+        requireNonNull(txn);
+        requireNonNull(config);
+        requireNonNull(recordBuilder);
+        requireNonNull(topicStore);
+
+        final var op = txn.getConsensusSubmitMessage();
+        final var topic = topicStore.getForModify(op.getTopicID().getTopicNum());
+        /* Validate all needed fields in the transaction */
+        validateTransaction(txn, config, topic);
+
+        /* since we have validated topic exists, topic.get() is safe to be called */
+        try {
+            final var updatedTopic = updateRunningHashAndSequenceNumber(txn, topic.get(), handleContext.consensusNow());
+            /* persist the updated topic */
+            topicStore.put(updatedTopic);
+        } catch (IOException e) {
+            throw new HandleStatusException(INVALID_TRANSACTION);
+        }
+    }
+
+    /**
+     * Validates te transaction body. Throws {@link HandleStatusException} if any of the validations fail.
+     * @param txn the {@link TransactionBody} of the active transaction
+     * @param config the {@link ConsensusServiceConfig}
+     * @param topic the topic to which the message is being submitted
+     */
+    private void validateTransaction(
+            final TransactionBody txn, final ConsensusServiceConfig config, final Optional<Topic> topic) {
+        final var txnId = txn.getTransactionID();
+        final var payer = txn.getTransactionID().getAccountID();
+        final var op = txn.getConsensusSubmitMessage();
+
+        /* Check if the message submitted is empty */
+        // Question do we need this check ?
+        if (op.getMessage().isEmpty()) {
+            throw new HandleStatusException(INVALID_TOPIC_MESSAGE);
+        }
+
+        /* Check if the message submitted is greater than acceptable size */
+        if (op.getMessage().size() > config.maxMessageSize()) {
+            throw new HandleStatusException(MESSAGE_SIZE_TOO_LARGE);
+        }
+
+        /* Check if the topic exists */
+        if (topic == null || topic.isEmpty()) {
```

Comment:
- First, I would avoid using `Optional` if I can because you still have to call `Optional.isEmpty()` before getting the instance, which is equivalent to checking if the instance is `null` itself.
- Second, also derived from the above, we should never check if the wrapper is `null`. If a method potentially returns a `null Optional wrapper` of a potentially `null` instance, that method needs to be refactored.

```suggestion
        if (topic.isEmpty()) {
```

Reply:
Either you have an Optional that can ever be null, or you don't have an Optional and the value can be null. But it doesn't make sense to have an Optional that can be null.

Reply:
An `Optional` that can be null doesn't make sense to me either, but java allows it 🤷 Especially for a method parameter I figured it's safer. I'm fine deferring though

Even better than `if (topic.isEmpty())`: `topic.orElseThrow()`

Reply:
Ah yes agreed. Removed `topic == null`